lib/Transforms/Functional.cpp

```diff
@@ -0,0 +1,153 @@
+//===- Functional.cpp -------------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "Transforms/Functional.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+
+using namespace mlir;
+using functional::impl::SequencePattern;
+
+Operation *mlir::functional::findNearestScope(Operation *op) {
+  do {
+    op = op->getParentOp();
+  } while (op && !op->hasTrait<OpTrait::IsIsolatedFromAbove>());
+  return op;
+}
+
+SmallVector<std::unique_ptr<Region>>
+mlir::functional::cloneRegionsOf(Operation *op) {
+  SmallVector<std::unique_ptr<Region>> regions;
+  BlockAndValueMapping mapping;
+  for (Region &region : op->getRegions()) {
+    region.cloneInto(&*regions.emplace_back(std::make_unique<Region>()),
+                     mapping);
+  }
+  return regions;
+}
+
+void mlir::functional::takeRegions(Operation *op,
+                                   ArrayRef<std::unique_ptr<Region>> regions) {
+  assert(op->getNumRegions() == regions.size() && "unequal number of regions");
+  for (auto it : llvm::zip(op->getRegions(), regions)) {
+    Region &region = std::get<0>(it);
+    region.getBlocks().clear();
+    region.takeBody(*std::get<1>(it));
+  }
+}
+
+namespace {
+class RecursiveMatchDriver : public PatternRewriter {
```

Comment:
Unless I missing something, it seems that everything below here is not exercised in the revision (or if it is is is hard for me to find)?

Could we split the recursive concepts it in a separate revision where we more easily see where they get applied ?

Reply:
Let me just remove this from the revision. I was experimenting with a lot of stuff on a separate branch and I just imported all the changes over.